src/wp-admin/includes/image.php

```diff
@@ -522,9 +522,25 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id, $mim
 				// within sources.
 				if ( $size_meta['mime-type'] === $mime_type || isset( $size_meta['sources'][ $mime_type ] ) ) {
 					unset( $new_sizes[ $size_name ] );
+					continue;
 				}
 			}
+
+			// Short-cicuit image generation if file has been generated already.
+			$size_meta = apply_filters( 'wp_content_pre_generate_additional_image_source', $size_meta, $file, $attachment_id, $size_name, $mime_type );
+
+			if ( is_wp_error( $size_meta ) ) {
```

Comment:
should this also skip if return is false?

Reply:
still suggest checking for false here @akshitsethi - if this used false a developer could do `add_filter( 'wp_content_pre_generate_additional_image_source', '__return_false' );` - why use `is_wp_error`?

Reply:
@adamsilverstein Here is the approach that I've implemented:

- The `$size_meta` in filter can be used to replace the filename which has either already being generated by the plugin or they can generate within the hook and update size meta value.
- The returned `$size_meta` value is checked to ensure that the mime source file for the size exists and if it does we skip the generation part.
- Once the generation for other sizes is done or is short-circuited, the attachment metadata is updated.

The approach you're suggesting is to pass a boolean to check for generation of the size wherein the plugins can call `'wp_content_pre_generate_additional_image_source', '__return_false' );` and skip the generation part.

I'm making changes to the code and tests to follow your approach but my only concern is that the structure we are trying to implement in core for different mime types might not be followed as it should be.

Reply:
> The $size_meta in filter can be used to replace the filename which has either already being generated by the plugin or they can generate within the hook and update size meta value.

OK, that makes sense. Can you clarify when `is_wp_error` return true here? Also please document the filter return (and how to use it)